cudapoa/src/cudapoa_kernels.cu

```diff
@@ -649,12 +685,75 @@ void generatePOA(claragenomics::cudapoa::OutputDetails* output_details_d,
+                 int8_t output_mask,
+                 const BatchSize& batch_size)
+{
+    // a decision flag to determine proper type definition for ScoreT
+    const bool use_32_bit_for_ScoreT = use32bitInt(batch_size.max_sequence_size, match_score);
+
+    if (use_32_bit_for_ScoreT)
+    {
+        claragenomics::cudapoa::AlignmentDetails<int32_t>* alignment_details_d = static_cast<claragenomics::cudapoa::AlignmentDetails<int32_t>*>(alignment_details_void);
+        claragenomics::cudapoa::generatePOAtemplated<int32_t>(output_details_d,
+                                                              input_details_d,
+                                                              total_windows,
+                                                              stream,
+                                                              alignment_details_d,
+                                                              graph_details_d,
+                                                              (int32_t)gap_score,
+                                                              (int32_t)mismatch_score,
+                                                              (int32_t)match_score,
+                                                              cuda_banded_alignment,
+                                                              max_sequences_per_poa,
+                                                              output_mask,
+                                                              batch_size);
+    }
+    else
+    {
+        claragenomics::cudapoa::AlignmentDetails<int16_t>* alignment_details_d = static_cast<claragenomics::cudapoa::AlignmentDetails<int16_t>*>(alignment_details_void);
+        claragenomics::cudapoa::generatePOAtemplated<int16_t>(output_details_d,
+                                                              input_details_d,
+                                                              total_windows,
+                                                              stream,
+                                                              alignment_details_d,
+                                                              graph_details_d,
+                                                              gap_score,
+                                                              mismatch_score,
+                                                              match_score,
+                                                              cuda_banded_alignment,
+                                                              max_sequences_per_poa,
+                                                              output_mask,
+                                                              batch_size);
+    }
+}
+
+bool use32bitInt(int32_t max_sequence_size, int32_t match_score)
+{
+    // a decision flag to determine proper type definition for ScoreT, factor 2 is selected ad-hoc
+    return (max_sequence_size * match_score * 2 > INT16_MAX);
```

Comment:
We can probably update the limits for this to be the following - 

max - `max_sequence_size * match_score`. This will only happen when sequence and graph match exactly. any other scenario will cause gap scores to be added which (as long as they're negative) will drive down the score.

min - no match whatsoever, so basically all gaps. and this would be `(max_sequence_size + max_graph_size) * gap_score`. 

what do you think?

Reply:
good point, I changed the condition to make decision based on both min and max, and removed the ad-hoc factor. may need to revise the min condition for banded.

Reply:
I replaced with the following logic which is more accurate and less conservative: if `d1 = max_sequence_size`, and `d2 = max_graph_size (assuming d2>=d1)`, `n = mismatch_score` and `g = gap_score`, where `n` and `g` are negative. Then minimum negative score computed in NW matrix should be bound to:
`d1*(max(g,n)) + (d2-d1)*g`

Reply:
Looks good to me! Can you add a comment explaining the min and max score calculations in the function?

Reply:
done.